lib/fluent/plugin/out_papertrail.rb

```diff
@@ -1,30 +1,84 @@
-require 'syslog_protocol'
-require 'fluent/output'
-
 module Fluent
-  class Papertrail < Fluent::Output
-    attr_accessor :socket, :papertrail_host, :papertrail_port
+  class Papertrail < Fluent::BufferedOutput
+    attr_accessor :socket
+
+    # if left empty in fluent config these config_param's will error
+    config_param :papertrail_host, :string
+    config_param :papertrail_port, :integer
+    # overriding default flush_interval (60 sec) from Fluent::BufferedOutput
+    config_param :flush_interval, :time, default: 1
 
+    # register as 'papertrail' fluent plugin
     Fluent::Plugin.register_output('papertrail', self)
 
+    def initialize
+      super
+      require 'syslog_protocol'
+    end
+
     def configure(conf)
       super
-      @papertrail_host = conf["papertrail_host"]
-      @papertrail_port = conf["papertrail_port"]
-      @socket = UDPSocket.new
+      @socket = create_socket(@papertrail_host, @papertrail_port)
+    end
+
+    def format(tag, time, record)
+      [tag, time, record].to_msgpack
+    end
+
+    def write(chunk)
+      chunk.msgpack_each {|(tag, time, record)|
+        packet = create_packet(tag, time, record)
+        send_to_papertrail(packet)
+      }
+    end
+
+    def create_socket(host, port)
+      log.info "initializing tcp socket for #{host}:#{port}"
+      begin
+        socket = TCPSocket.new(host, port)
+        log.debug "enabling ssl for socket #{host}:#{port}"
+        ssl = OpenSSL::SSL::SSLSocket.new(socket)
+        # close tcp and ssl socket when either fails
+        ssl.sync_close = true
+        # initiate SSL/TLS handshake with server
+        ssl.connect
+      rescue => e
+        log.warn "failed to create tcp socket #{host}:#{port}: #{e}"
+        ssl = nil
+      end
+      ssl
     end
 
-    def emit(tag, es, chain)
-      es.each do |time, record|
-        packet = SyslogProtocol::Packet.new
-        packet.hostname = record['hostname']
-        packet.facility = record['facility']
-        packet.severity = record['severity']
-        packet.tag = record['program']
-        packet.content = record['message']
-        @socket.send(packet.assemble, 0, papertrail_host, papertrail_port)
+    def create_packet(tag,time,record)
+      # construct syslog packet from fluent record
+      packet = SyslogProtocol::Packet.new
+      packet.hostname = record['hostname']
+      packet.facility = record['facility']
+      packet.severity = record['severity']
+      packet.tag      = record['program']
+      packet.content  = record['message']
+      packet.time     = time ? Time.at(time) : Time.now
+      packet
+    end
+
+    def send_to_papertrail(packet)
+      # recreate the socket if it's nil -- see below
+      unless @socket
+        @socket = create_socket(@papertrail_host, @papertrail_port)
+      end
+
+      if @socket
```

Comment:
I'd rather see `if !@socket.nil?` b/c Ruby is loosely typed and there are a lot of falsey values. I think this is just several years of Go talking here tho, so do with that what you will.

Reply:
No, that makes sense. Will do. Seems to be idiomatic as well. Ruby Style Guide says to use `unless` instead of `if !`, but not to use `unless` with `else`, so I'll reorder this statement.

Reply:
Yeah `unless...else` tends to tie people's heads in knots. I use `unless` mostly with one-liners for flow control.